File: samples/SampleIssueTracker/Issues/Domain/IssueSnapshot.cs
using SampleIssueTracker.SharedKernel;

namespace SampleIssueTracker.Issues.Domain;

/// <summary>
/// Flat JSON projection of an <see cref="Issue"/> plus its comments, designed to
/// round-trip through Metano's <c>JsonSerializer</c>.
///
/// <para>
/// Lives alongside the aggregate in the domain package so the wire shape stays
/// close to its source of truth, but is kept as a separate type because
/// <see cref="Issue"/> itself has encapsulated state (private setters,
/// state-mutating methods) that does not round-trip cleanly from a flat factory
/// constructor. This is the canonical "snapshot for serialization" pattern —
/// rich aggregates expose a <c>ToSnapshot()</c> / <c>FromSnapshot()</c> pair and
/// the wire format operates on the snapshot, not the aggregate itself.
/// </para>
///
/// <para>
/// Also the only type in the sample that carries a <see cref="decimal"/>, which
/// exists so the Phase 5 serialization tests can exercise the custom-converter
/// code path (decimal → JSON number by default, decimal → JSON string when a
/// context-level converter is registered).
/// </para>
/// </summary>
public record IssueSnapshot(
    IssueId Id,
    string Title,
    string Description,
    IssueType Type,
    IssuePriority Priority,
    IssueStatus Status,
    UserId? AssigneeId,
    string? SprintKey,
    DateTimeOffset CreatedAt,
    DateTimeOffset UpdatedAt,
    decimal EstimatedHours,
    Comment[] Comments
);

File: js/sample-issue-tracker/src/issues/domain/issue-snapshot.ts
import { HashCode } from "metano-runtime";
import { Temporal } from "@js-temporal/polyfill";
import { Decimal } from "decimal.js";
import type { UserId } from "#/shared-kernel";
import type { Comment } from "./comment";
import type { IssueId } from "./issue-id";
import { IssuePriority } from "./issue-priority";
import { IssueStatus } from "./issue-status";
import { IssueType } from "./issue-type";

export class IssueSnapshot {
  constructor(readonly id: IssueId, readonly title: string, readonly description: string, readonly type: IssueType, readonly priority: IssuePriority, readonly status: IssueStatus, readonly assigneeId: UserId | null, readonly sprintKey: string | null, readonly createdAt: Temporal.ZonedDateTime, readonly updatedAt: Temporal.ZonedDateTime, readonly estimatedHours: Decimal, readonly comments: Comment[]) { }

  equals(other: any): boolean {
    return other instanceof IssueSnapshot && this.id === other.id && this.title === other.title && this.description === other.description && this.type === other.type && this.priority === other.priority && this.status === other.status && this.assigneeId === other.assigneeId && this.sprintKey === other.sprintKey && this.createdAt === other.createdAt && this.updatedAt === other.updatedAt && this.estimatedHours === other.estimatedHours && this.comments === other.comments;
  }

  hashCode(): number {
    const hc = new HashCode();
    hc.add(this.id);
    hc.add(this.title);
    hc.add(this.description);
    hc.add(this.type);
    hc.add(this.priority);
    hc.add(this.status);
    hc.add(this.assigneeId);
    hc.add(this.sprintKey);
    hc.add(this.createdAt);
    hc.add(this.updatedAt);
    hc.add(this.estimatedHours);
    hc.add(this.comments);
    return hc.toHashCode();
  }

  with(overrides?: Partial<IssueSnapshot>): IssueSnapshot {
    return new IssueSnapshot(overrides?.id ?? this.id, overrides?.title ?? this.title, overrides?.description ?? this.description, overrides?.type ?? this.type, overrides?.priority ?? this.priority, overrides?.status ?? this.status, overrides?.assigneeId ?? this.assigneeId, overrides?.sprintKey ?? this.sprintKey, overrides?.createdAt ?? this.createdAt, overrides?.updatedAt ?? this.updatedAt, overrides?.estimatedHours ?? this.estimatedHours, overrides?.comments ?? this.comments);
  }
}
